packages/desktop-client/src/components/reports/reports/CustomReport.js

```diff
@@ -102,6 +101,13 @@ export default function CustomReport() {
     }
     run();
   }, []);
+
+  const OnChangeLegend = legend => {
+    useEffect(() => {
+      setLegendData(legend);
+    }, []);
+  };
```

Comment:
<!-- Warning represent user-facing problems. If possible, it's great to follow this kind of comment with a suggestion.  -->
**🔨 warning:** ‏we can definitely improve this. The usage of `useEffect` seems to be redundant here unless I'm missing something.

```suggestion
  const OnChangeLegend = setLegendData;
```

Reply:
Without the use of "useEffect" or "useMemo" or something similar we end up with an infinite loop becuase of the way recharts handles passing legend data via payloads.

As a result of using a hook here the eslint rules throw an error about camelCase and request that the element start with a capital letter.

Reply:
Looked into this more deeper.

The root-cause issue here is [two-way data flow](https://angular.io/guide/two-way-binding) in your components. I've seen this in Angular, but never in React-based projects. React isn't really built around two-way data bindings. The data-flow here is unidirectional (for the most part).

So the problem is that you are calculating the color-map on the `*Chart` (child) components and then trying to figure out how to pass these calculated colors to the root (consumer) component. Instead you can calculate the colors on root-level thus removing the need to pass this data back and forth.

So instead of having the color calculation [here](https://github.com/actualbudget/actual/blob/8a15caf42dc4cfecd6b82f24712e90c914d385c2/packages/desktop-client/src/components/reports/graphs/BarGraph.tsx#L221-L227) you would perform the calculation operation [here](https://github.com/actualbudget/actual/blob/8a15caf42dc4cfecd6b82f24712e90c914d385c2/packages/desktop-client/src/components/reports/reports/CustomReport.js#L166-L169). Which means you also get rid of the `CustomLabel` component that doesn't really do anything. Plus you get rid of all the `OnChangeLegend` callbacks and their prop drilling.

Reply:
Huge thanks mate. I always felt that "custom legend" in the graph files felt a bit hacky. I was happy I finally got it workting but it was all a bit backwards as you say. Just need some outside eyes on it to find where I went wrong. Sometimes I spend so much time looking at the same code and my brain stops working.

New version brings all the color calcs into the spreadsheet so they are consistent across all graphs and all in one place. The calculation is a bit complex but not too bad IMO (considering it is trying to set colors for 3 different graphs, all with different needs). LMK what you think.